File: src/main/java/com/williambl/haema/vampire/ability/powers/sunlight_sickness/SunlightSicknessEffect.java
package com.williambl.haema.vampire.ability.powers.sunlight_sickness;

import com.williambl.haema.api.vampire.VampireComponent;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.effect.MobEffectCategory;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.ai.attributes.AttributeModifier;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.NotNull;

public class SunlightSicknessEffect extends MobEffect {
    public SunlightSicknessEffect() {
        super(MobEffectCategory.HARMFUL, 0xf5a74200);
        this.addAttributeModifier(
                Attributes.ATTACK_DAMAGE,
                "c85d1cfe-2c10-4d25-b650-49c045979842",
                -4.0,
                AttributeModifier.Operation.ADDITION);
    }

    @Override
    public void applyEffectTick(@NotNull LivingEntity entity, int amplifier) {
        if (entity instanceof Player p && p.isCreative()) {
            return;
        }

        var component = VampireComponent.KEY.getNullable(entity);
        if (component == null) {
            return;
        }

        if (entity.tickCount % 10 == 0) {
            entity.hurt(entity.damageSources().inFire(), 0.5f + amplifier); //TODO damage source
            if (component.isVampire()) {
                component.setBlood(component.getBlood() - 0.25);
            }
            if (entity.level() instanceof ServerLevel level) {
                level.sendParticles(
                        ParticleTypes.FLAME,
                        entity.getX() - entity.getBbWidth() / 4.0,
                        entity.getY(),
                        entity.getZ() - entity.getBbWidth() / 4.0,
                        10,
                        entity.getBbWidth()/2.0,
                        entity.getBbHeight()/2.0,
                        entity.getBbWidth()/2.0,
                        0.0
                );
            }
        }
    }

    @Override
    public boolean shouldApplyEffectTickThisTick(int duration, int amplifier) {
        return true;
    }
}

File: src/main/kotlin/com/williambl/haema/effect/SunlightSicknessEffect.kt
package com.williambl.haema.effect

import com.williambl.haema.damagesource.SunlightDamageSource
import com.williambl.haema.isVampire
import com.williambl.haema.vampireComponent
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.attribute.EntityAttributeModifier
import net.minecraft.entity.attribute.EntityAttributes
import net.minecraft.entity.effect.StatusEffect
import net.minecraft.entity.effect.StatusEffectCategory
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.particle.ParticleTypes
import net.minecraft.server.world.ServerWorld

class SunlightSicknessEffect(type: StatusEffectCategory?, color: Int) : StatusEffect(type, color) {

    companion object {
        val instance: StatusEffect = SunlightSicknessEffect(StatusEffectCategory.HARMFUL, 245 shl 24 or 167 shl 16 or 66 shl 8)
                .addAttributeModifier(
                        EntityAttributes.GENERIC_ATTACK_DAMAGE,
                        "c85d1cfe-2c10-4d25-b650-49c045979842",
                        -4.0,
                        EntityAttributeModifier.Operation.ADDITION
                )
    }

    override fun applyUpdateEffect(entity: LivingEntity?, amplifier: Int) {
        if (entity == null || !entity.isVampire || entity.isSpectator || (entity is PlayerEntity && entity.isCreative)) {
            return
        }

        if (entity.age % 10 == 0) {
            entity.damage(SunlightDamageSource.instance, 0.5f)
            if (entity.isVampire) {
                (entity.vampireComponent).removeBlood(0.25)
            }
            val pos = entity.pos
            val world = entity.world
            if (world is ServerWorld) {
                world.spawnParticles(
                    ParticleTypes.FLAME,
                    pos.x - entity.width/4.0,
                    pos.y,
                    pos.z - entity.width/4.0,
                    10,
                    entity.width.toDouble()/2.0,
                    entity.height.toDouble()/2.0,
                    entity.width.toDouble()/2.0,
                    0.0
                )
            }
        }
    }

    override fun canApplyUpdateEffect(duration: Int, amplifier: Int): Boolean {
        return true
    }
}